src/app/clusters/scenes/ExtensionFieldsSetsImpl.h

```diff
@@ -0,0 +1,159 @@
+/**
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/clusters/scenes/ExtensionFieldsSets.h>
+#include <lib/core/DataModelTypes.h>
+
+namespace chip {
+namespace scenes {
+
+using clusterId = chip::ClusterId;
+
+struct ExtensionFieldsSet
+{
+    static constexpr TLV::Tag TagClusterId() { return TLV::ContextTag(1); }
+    static constexpr TLV::Tag TagEFS() { return TLV::ContextTag(2); }
+    clusterId mID                              = kInvalidClusterId;
+    uint8_t mBytesBuffer[kMaxFieldsPerCluster] = { 0 };
+    uint8_t mUsedBytes                         = 0;
+
+    ExtensionFieldsSet() = default;
+    ExtensionFieldsSet(clusterId cmID, uint8_t * data, uint8_t dataSize) : mID(cmID), mUsedBytes(dataSize)
+    {
+        if (dataSize <= kMaxFieldsPerCluster)
+        {
+            memcpy(mBytesBuffer, data, mUsedBytes);
+        }
+    }
+    ~ExtensionFieldsSet() = default;
+
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const
+    {
+        TLV::TLVType container;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::ContextTag(1), TLV::kTLVType_Structure, container));
+
+        ReturnErrorOnFailure(writer.Put(TagClusterId(), static_cast<uint16_t>(this->mID)));
+        ReturnErrorOnFailure(writer.PutBytes(TagEFS(), mBytesBuffer, mUsedBytes));
+
+        return writer.EndContainer(container);
+    }
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader)
+    {
+        ByteSpan buffer;
+        TLV::TLVType container;
+        ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::ContextTag(1)));
+        ReturnErrorOnFailure(reader.EnterContainer(container));
+
+        ReturnErrorOnFailure(reader.Next(TagClusterId()));
+        ReturnErrorOnFailure(reader.Get(this->mID));
+
+        ReturnErrorOnFailure(reader.Next(TagEFS()));
+        ReturnErrorOnFailure(reader.Get(buffer));
+        if (buffer.size() > kMaxFieldsPerCluster)
+        {
+            this->mUsedBytes = kMaxFieldsPerCluster;
```

Comment:
We cannot and should not truncate here.

Reply:
Now returns error if size is too big instead of truncating.